File: Sources/SomnioCore/Placement/VisualCenter.swift
import Foundation

/// Visual-center geometry helpers for radius gates that compare entity centers, not their
/// top-left origins. Used by the AI tick to decide whether a player is within an NPC's
/// dialog radius (`SomnioConstants.npcInteractionRadius`) or a monster's aggro radius
/// (`SomnioConstants.monsterAggroRadius`). Lives next to `NPCPlacement` because both
/// helpers materialize runtime geometry the codec deliberately keeps out.
///
/// Centers are passed in as `Int32` pairs and squared distances accumulate in `Int64` so a
/// player at the edge of a sector with `dimensions.width` near `Int16.max` cannot trap the
/// AI tick on the `dx*dx + dy*dy` square.
public enum VisualCenter {
    public static func squaredDistance(_ a: (x: Int32, y: Int32), _ b: (x: Int32, y: Int32)) -> Int64 {
        let dx = Int64(a.x) - Int64(b.x)
        let dy = Int64(a.y) - Int64(b.y)
        return dx * dx + dy * dy
    }

    /// Inclusive radius gate: a point exactly at the radius is inside.
    public static func isWithin(_ a: (x: Int32, y: Int32), _ b: (x: Int32, y: Int32), radius: Int16) -> Bool {
        let r = Int64(radius)
        return squaredDistance(a, b) <= r * r
    }
}

File: packages/core/src/visualCenter.ts
/**
 * Visual-center geometry for the radius gates that compare entity centers, not their top-left
 * origins: the NPC dialog radius and the monster aggro radius. Squared distances stay within the
 * safe-integer range for any pair of Int16 centers, so a sector edge near `Int16.max` cannot
 * produce a wrong verdict.
 */
export interface Center {
  x: number;
  y: number;
}

export function squaredDistance(a: Center, b: Center): number {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  return dx * dx + dy * dy;
}

/** Inclusive radius gate: a point exactly at the radius is inside. */
export function isWithin(a: Center, b: Center, radius: number): boolean {
  return squaredDistance(a, b) <= radius * radius;
}
